x-pack/plugins/security_solution/server/lib/detection_engine/rule_response_actions/endpoint_response_action.ts

```diff
@@ -5,81 +5,149 @@
-import { isIsolateAction, isProcessesAction } from './endpoint_params_type_guards';
-import type { RuleResponseEndpointAction } from '../../../../common/api/detection_engine';
+import type {
+  RuleResponseEndpointAction,
+  ProcessesParams,
+} from '../../../../common/api/detection_engine';
+import type { KillOrSuspendProcessRequestBody } from '../../../../common/endpoint/types';
+import { getErrorProcessAlerts, getIsolateAlerts, getProcessAlerts } from './utils';
+import type { AlertsAction, ResponseActionAlerts } from './types';
 import type { EndpointAppContextService } from '../../../endpoint/endpoint_app_context_services';
-import { getProcessAlerts, getIsolateAlerts, getErrorProcessAlerts } from './utils';
 
-import type { ResponseActionAlerts, AlertsAction } from './types';
+const NOOP = () => {};
 
-export const endpointResponseAction = (
+export const endpointResponseAction = async (
   responseAction: RuleResponseEndpointAction,
   endpointAppContextService: EndpointAppContextService,
-  { alerts }: ResponseActionAlerts,
-  experimentalFeatures: ExperimentalFeatures
-) => {
+  { alerts }: ResponseActionAlerts
+): Promise<void> => {
+  const logger = endpointAppContextService.createLogger(
+    'ruleExecution',
+    'automatedResponseActions'
+  );
+  const ruleId = alerts[0][ALERT_RULE_UUID];
+  const ruleName = alerts[0][ALERT_RULE_NAME];
+  const logMsgPrefix = `Rule [${ruleName}][${ruleId}]:`;
   const { comment, command } = responseAction.params;
+  const agentIds = uniq(map(alerts, 'agent.id'));
+  const errors: string[] = [];
+  const responseActionsClient = endpointAppContextService.getInternalResponseActionsClient({
+    agentType: 'endpoint',
+    username: 'unknown',
+  });
+
+  const automatedProcessActionsEnabled =
+    endpointAppContextService.experimentalFeatures.automatedProcessActionsEnabled;
+
+  const processResponseActionClientError = (err: Error, endpointIds: string[]): Promise<void> => {
+    errors.push(
+      `attempt to isolate host [${endpointIds.join(', ')}] returned error: ${err.message}`
+    );
 
-  const commonData = {
-    comment,
-    command,
-    rule_id: alerts[0][ALERT_RULE_UUID],
-    rule_name: alerts[0][ALERT_RULE_NAME],
-    agent_type: 'endpoint' as const,
+    return Promise.resolve();
   };
 
-  if (isIsolateAction(responseAction.params)) {
-    const alertsPerAgent = getIsolateAlerts(alerts);
-    each(alertsPerAgent, (actionPayload) => {
-      return endpointAppContextService.getActionCreateService().createActionFromAlert(
-        {
-          ...actionPayload,
-          ...commonData,
-        },
-        actionPayload.endpoint_ids
-      );
-    });
-  }
+  logger.info(`${logMsgPrefix} will create a total of [${agentIds.length}] [${command}] action(s)`);
+
+  let response: Promise<void> = Promise.resolve();
+
+  switch (command) {
+    case 'isolate':
+      response = Promise.all(
+        // TODO:PT should this be throttled?
```

Comment:
Good question! Have we faced performance issues regarding isolate anytime before? In this situation we will send one action per host so I don't think we should worry about it, but... would love to hear your thoughts. 
Regarding other actions like kill-process, where we might send multiple actions to one host - I have some trust in analysts that create the rules :P The rules tend to be very very specific, meaning shouldn't really trigger many actions at a time. 
Anyway, what kind of throttling do you suggest?

Reply:
When I looked at this code, my thought was: what happens if there 10k alerts for 10 unique agents and they all send 3 actions? Will that overload ES/KB as they are all trying to be done concurrently?

The throttling would be to perhaps only (example) no more than 10 at a time - by using this: 

https://github.com/elastic/kibana/blob/9150f9fa2f110bcd54f90f21554bad5e6d92fd0f/x-pack/plugins/security_solution/server/endpoint/utils/queue_processor.ts#L21-L43

Reply:
Not sure about the performance ;/ We have osquery action prepared in a way that it is called 1 action per 1 alert (if there is a dynamic value included). That hasn't exploded yet :P 

We try to calculate and limit the number of actions to the bear minimum, so we do not send 2 exactly same actions. 
I don't think we will create sooo many actions, even when testing I used 2-3 agents with very broad rule query (*:*) that triggered hundreds of kill-process actions. Not sure what the real numbers are for our customers, but I would assume that we will see the same limitations as anywhere when it comes to fleet limitations. 
But of course, worth checking 👍 And also throttling would not hurt us I guess
What do you think?